fda/registration.py

```diff
@@ -0,0 +1,490 @@
+        ..  [RGS09-8-5] Ramsay J.O., Giles Hooker & Spencer Graves (2009). In
+            *Functional Data Analysis with R and Matlab* (pp. 125-126).
+            Springer.
+    """
+
+    if original_fdata.nsamples != registered_fdata.nsamples:
+        raise ValueError(f"the registered and unregistered curves must have "
+                         f"the same number of samples "
+                         f"({registered_fdata.nsamples})!= "
+                         f"({original_fdata.nsamples})")
+
+    if warping_function is not None and (warping_function.nsamples
+                                         != original_fdata.nsamples):
+        raise ValueError(f"the registered curves and the warping functions must"
+                         f" have the same number of samples "
+                         f"({registered_fdata.nsamples})"
+                         f"!=({warping_function.nsamples})")
+
+    # Creates the mesh to discretize the functions
+    if tfine is None:
+        nfine = max(registered_fdata.basis.nbasis * 10 + 1, 201)
+        tfine = numpy.linspace(*registered_fdata.domain_range, nfine)
+    else:
+        tfine = numpy.asarray(tfine)
+
+    x_fine = original_fdata.evaluate(tfine) # Unregistered function
+    y_fine = registered_fdata.evaluate(tfine) # Registered function
+    mu_fine = x_fine.mean(axis=0) # Mean unregistered function
+    eta_fine = y_fine.mean(axis=0) # Mean registered function
+    mu_fine_sq = numpy.square(mu_fine)
+    eta_fine_sq = numpy.square(eta_fine)
+
+
+    # Total mean square error of the original funtions
+    mse_total = scipy.integrate.simps(
+        numpy.mean(numpy.square(x_fine - mu_fine), axis=0),tfine)
+
+    cr = 1. # Constant related to the covariation between the deformation
+            # functions and y^2
+
+    # If the warping functions are not provided, are suppose to be independent
+    if warping_function is not None:
+        # Derivates warping functions
+        dh_fine = warping_function.evaluate(tfine, derivative=1)
+        dh_fine_mean = dh_fine.mean(axis=0)
+        dh_fine_center = dh_fine - dh_fine_mean
+
+        y_fine_sq = numpy.square(y_fine) # y^2
+        y_fine_sq_center = numpy.subtract(y_fine_sq, eta_fine_sq) # y^2 - E[y^2]
+
+        covariate = numpy.inner(dh_fine_center.T, y_fine_sq_center.T)
+        covariate = covariate.mean(axis=0)
+        cr += numpy.divide(scipy.integrate.simps(covariate, tfine),
+                           scipy.integrate.simps(eta_fine_sq, tfine))
+
+
+    # mse due to phase variation
+    mse_pha = scipy.integrate.simps(cr*eta_fine_sq - mu_fine_sq , tfine)
+
+    # mse due to amplitude variation
+    mse_amp = mse_total - mse_pha
+
+    # squared correlation measure of proportion of phase variation
+    rsq = mse_pha / (mse_total)
+
+    return mse_amp, mse_pha, rsq, cr
+
+
+
+def shift_registration_deltas(fd, maxiter=5, tol=1e-2, extrapolation=None,
+                              step_size=1, initial=None, tfine=None, **kwargs):
```

Comment:
I am still doubting if `tfine` is a good name. What is the reason for choosing it?